src/test/groovy/graphql/validation/rules/NoUndefinedVariablesTest.groovy

```diff
@@ -181,4 +181,37 @@ class NoUndefinedVariablesTest extends Specification {
         errorCollector.containsValidationError(ValidationErrorType.UndefinedVariable)
         errorCollector.getErrors()[0].message == "Validation error (UndefinedVariable@[A/field]) : Undefined variable 'a'"
     }
-}
+
+    def "variable-directive back-reference"() {
+        given:
+        def query = "query (\$v1: Int, \$v2: Int @dir(arg: \$v1)) { __typename }"
+
+        when:
+        traverse(query)
+
+        then:
+        errorCollector.errors.isEmpty()
+    }
+
+    def "variable-directive self-reference"() {
+        given:
+        def query = "query (\$v1: Int @dir(arg: \$v1)) { __typename }"
+
+        when:
+        traverse(query)
+
+        then:
+        errorCollector.errors.isEmpty()
+    }
+
+    def "variable-directive forward-reference"() {
+        given:
+        def query = "query (\$v1: Int @dir(arg: \$v2), \$v2: Int) { __typename }"
```

Comment:
the other test cases in this file pass, though this one fails:
```
Undefined variable 'v2'
```

I think (but am not certain) that this pattern is allowed by the graphql spec.